usr/man/man3/monitor.3
.TH monitor 3 
.SH NAME
monitor, monstartup, moncontrol \- prepare execution profile
.SH SYNTAX
.nf
.B monitor(lowpc, highpc, buffer, bufsize, nfunc)
.B int (*lowpc)(), (*highpc)();
.B short buffer[];
.sp
.B monstartup(lowpc, highpc)
.B int (*lowpc)(), (*highpc)();
.sp
.B moncontrol(mode)
.fi
.SH DESCRIPTION
There are two different forms of monitoring available:
An executable program created by:
.IP  "	cc \-p . . ."
.LP
automatically includes calls for the prof(1)
monitor and includes an initial call to its start-up routine
.PN monstartup
with default parameters;
.PN monitor
need not be called explicitly except to gain fine control over profil 
buffer allocation.
An executable program created by:
.IP  "	cc \-pg . . ."
.LP
automatically includes calls for the gprof(1) monitor.
.PP
The
.PN monstartup
is a high level interface to profil(2). The
.I lowpc
and
.I highpc
specify the address range that is to be sampled; the lowest address sampled
is that of
.I lowpc
and the highest is just below
.IR highpc .
The
.PN monstartup
subroutine allocates space using sbrk(2)
and passes it to
.PN monitor
(see below) to record a histogram of periodically sampled values of
the program counter, and of counts of calls of certain functions, in the buffer.
Only calls of functions compiled with the profiling option
.B \-p
of cc(1) are recorded.
.PP
To profile the entire program, it is sufficient to use
.EX
extern etext();
. . .
monstartup((int) 2, etext);
.EE
The
.I etext
lies just above all the program text, see end(3).
.PP
To stop execution monitoring and write the results on the file
mon.out, use
.EX
monitor(0);
.EE
then prof(1) can be used to examine the results.
.PP
The
.PN moncontrol
subroutine is used to selectively control profiling within a program.
This works with either prof(1) or gprof(1)
type profiling.
When the program starts, profiling begins.
To stop the collection of histogram ticks and call counts use
moncontrol(0);
to resume the collection of histogram ticks and call counts use
moncontrol(1).
This allows the cost of particular operations to be measured.
Note that an output file will be produced upon program exit
regardless of the state of 
.PN moncontrol.
.PP
The
.PN monitor
subroutine
is a low level interface to profil(2).
The
.I lowpc
and
.I highpc
are the addresses of two functions;
.I buffer
is the address of a (user supplied) array of
.I bufsize
short integers.  At most
.I nfunc
call counts can be kept.
For the results to be significant,
especially where there are small, heavily
used routines, it is suggested that the buffer be no more
than a few times smaller than the range of locations sampled.
The
.PN monitor
subroutine
divides the buffer into space to record the histogram
of program counter samples over the range
.I lowpc
to
.IR highpc ,
and space to record call counts of functions compiled with the 
.B \-p
option to cc(1).
.PP
To profile the entire program, it is sufficient to use
.EX
extern etext();
. . .
monitor((int) 2, etext, buf, bufsize, nfunc);
.EE
.SH FILES
mon.out
.SH "SEE ALSO"
cc(1), gprof(1), prof(1), profil(2), sbrk(2)


new/ingres/source/monitor/monitor.c
#ifndef lint
static	char	*sccsid = "@(#)monitor.c	1.1	(ULTRIX)	1/8/85";
#endif lint

/************************************************************************
 *									*
 *			Copyright (c) 1984 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/

# include	"monitor.h"
# include	<ingres.h>
# include	<aux.h>
# include	<signal.h>
# include	<ctlmod.h>




/*
**  MONITOR
**
**	This routine maintains the logical query buffer in
**	/tmp/INGQxxxx.  It in general just does a copy from input
**	to query buffer, unless it gets a backslash escape character
**	or dollarsign escape character.
**	It recognizes the following escapes:
**
**	\a -- force append mode (no autoclear)
**	\b -- branch (within an include file only)
**	\c -- reserved for screen clear in geoquel
**	\d -- change working directory
**	\e -- enter editor
**	\g -- "GO": submit query to INGRES
**	\i -- include (switch input to external file)
**	\k -- mark (for \b)
**	\l -- list: print query buffer after macro evaluation
**	\p -- print query buffer (before macro evaluation)
**	\q -- quit ingres
**	\r -- force reset (clear) of query buffer
**	\s -- call shell
**	\t -- print current time
**	\v -- evaluate macros, but throw away result (for side effects)
**	\w -- write query buffer to external file
**	\$t -- change setting of trace flags
**	\$r -- reset system
**	\\ -- produce a single backslash in query buffer
**
**	Uses trace flag 2
*/

/*
**  COMMAND TABLE
**	To add synonyms for commands, add entries to this table
*/

struct cntrlwd
{
	char	*name;
	int	code;
};

struct cntrlwd	Controlwords[] =
{
	"a",		C_APPEND,
	"append",	C_APPEND,
	"b",		C_BRANCH,
	"branch",	C_BRANCH,
	"cd",		C_CHDIR,
	"chdir",	C_CHDIR,
	"e",		C_EDIT,
	"ed",		C_EDIT,
	"edit",		C_EDIT,
	"editor",	C_EDIT,
	"g",		C_GO,
	"go",		C_GO,
	"i",		C_INCLUDE,
	"include",	C_INCLUDE,
	"read",		C_INCLUDE,
	"k",		C_MARK,
	"mark",		C_MARK,
	"l",		C_LIST,
	"list",		C_LIST,
	"p",		C_PRINT,
	"print",	C_PRINT,
	"q",		C_QUIT,
	"quit",		C_QUIT,
	"r",		C_RESET,
	"reset",	C_RESET,
	"s",		C_SHELL,
	"sh",		C_SHELL,
	"shell",	C_SHELL,
	"t",		C_TIME,
	"time",		C_TIME,
	"date",		C_TIME,
	"v",		C_EVAL,
	"eval",		C_EVAL,
	"w",		C_WRITE,
	"write",	C_WRITE,
	"$t",		C_SYSTRACE,
	"$trace",	C_SYSTRACE,
	"$r",		C_SYSRESET,
	"$reset",	C_SYSRESET,
	0
};


monitor()
{
	register char	chr;
	int		timevec[2];
	register int	controlno;
	extern jmp_buf	CmReset;
	extern		error();
	extern char	*Proc_name;
	extern int	RubLevel;
	extern		rubcatch();

	setjmp(CmReset);
	initbuf(Qbuf, QbufSize, ERR_QBUF, error);
	clrmem(&Ctx, sizeof Ctx);
	Ctx.ctx_cmark = Ctx.ctx_pmark = markbuf(Qbuf);
	Ctx.ctx_name = Proc_name = Cm.cm_myname;
	Ctx.ctx_tvect = tT = FuncVect[0]->fn_tvect;
	xwait();
	if (RubLevel >= 0)
		signal(SIGINT, rubcatch);

	while (chr = getch())
	{
		if (chr == '\\')
		{
			/* process control sequence */
			if ((controlno = getescape(1)) == 0)
				continue;

			switch (controlno)
			{

			  case C_EDIT:
				edit();
				continue;

			  case C_PRINT:
				print();
				continue;

			  case C_LIST:
				eval(1);
				continue;

			  case C_EVAL:
				eval(0);
				Autoclear = TRUE;
				continue;

			  case C_INCLUDE:
				include(0);
				cgprompt();
				continue;

			  case C_WRITE:
				writeout();
				cgprompt();
				continue;

			  case C_CHDIR:
				newdirec();
				cgprompt();
				continue;

			  case C_RESET:
				clear(1);
				continue;

			  case C_GO:
				go();
				continue;

			  case C_QUIT:
				clrline(1);
				quit();

			  case C_SHELL:
				shell();
				continue;

			  case C_TIME:
				time(timevec);
				printf("%s", ctime(timevec));
				clrline(0);
				continue;

			  case C_APPEND:
				Autoclear = 0;
				clrline(0);
				continue;

			  case C_MARK:
				getfilenm();
				prompt(0);
				continue;

			  case C_BRANCH:
				branch();
				prompt(0);
				continue;

			  case C_SYSTRACE:
				trace();
				prompt(0);
				continue;

			  case C_SYSRESET:
				reset();
				clrline(0);
				continue;

			  default:
				syserr("monitor: bad code %d", controlno);
			}
		}
		putch(chr);
	}

	if (Input == stdin)
	{
		if (Nodayfile >= 0)
			printf("\n");
	}
	else
		fclose(Input);
}

getescape(copy)
int	copy;
{
	register struct cntrlwd	*cw;
	register char		*word;
	char			*getname();

	word = getname();
	for (cw = Controlwords; cw->name; cw++)
	{
		if (sequal(cw->name, word))
			return (cw->code);
	}

	/* not found -- pass symbol through and return failure */
	if (copy == 0)
		return (0);
	putch('\\');
	while (*word != 0)
	{
		putch(*word++);
	}
	return (0);
}

char *
getname()
{
	register char	*p;
	static char	buf[41];
	register int	len;
	register char	c;

	p = buf;
	for (len = 0; len < 40; len++)
	{
		c = getch();
		if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))
		{
			*p++ = c;
		}
		else if ((len == 0) && (c == '$'))	/* system control command */
		{
			*p++ = c;
		}
		else
		{
			ungetc(c, Input);
			break;
		}
	}

	*p = 0;
	return (buf);
}



putch(ch)
char	ch;
{
	register char	c;

	c = ch;

	Prompt = Newline = (c == '\n');
	if (c < 040 && c != '\n' && c != '\t')
	{
		printf("Funny character 0%o converted to blank\n", c);
		c = ' ';
	}
	prompt(0);
	if (Autoclear)
		clear(0);
	putc(c, Qryiop);
	Notnull++;
}
